pipeline/utils/github_utils.py

```diff
@@ -0,0 +1,137 @@
+    path_content = base64.b64encode(path_content)
+    blob = _get_request_content(requests.post,
+                                '{0}/repos/{1}/{2}'
+                                '/git/blobs'.format(_API_URL, owner, repo),
+                                json.dumps({'content': path_content,
+                                            'encoding': 'base64'}),
+                                username, password)
+    return {'path': file_or_exec,
+            'mode': _TREE_EL_MODES[blob_mode],
+            'type': 'blob',
+            'sha': blob['sha']}
+
+
+# TODO: Have this function also upload symlinks to the repository
+# according to GitHub protocol
+def _build_git_tree(curr_dir, username, password, owner, repo):
+    """Create a git tree from curr_dir for the given GitHub repo"""
+    # Tree_els is a list of children (trees or blobs) for this tree node.
+    tree_els = []
+    for subdir in _get_entries(os.path.isdir, curr_dir):
+        subtree = _build_git_tree(os.path.join(curr_dir, subdir), username,
+                                  password, owner, repo)
+        tree_els.append({'path': subdir,
+                         'mode': _TREE_EL_MODES['tree'],
+                         'type': 'tree', 'sha': subtree['sha']})
+    for afile in _get_entries(_is_file, curr_dir):
+        tree_els.append(_build_git_tree_blob_item(afile, 'file',
+                                                  curr_dir, username,
+                                                  password, owner,
+                                                  repo))
+    for executable in _get_entries(_is_executable, curr_dir):
+        tree_els.append(_build_git_tree_blob_item(executable,
+                                                  'executable',
+                                                  curr_dir, username,
+                                                  password, owner,
+                                                  repo))
+    return _get_request_content(requests.post,
+                                '{0}/repos/{1}/{2}/git/trees'.format(_API_URL,
+                                                                     owner,
+                                                                     repo),
+                                json.dumps({'tree': tree_els}), username,
+                                password)
+
+
+def push_dir_github(output_dir, username, password, owner, repo, branch,
+                    message='Automated commit from artman'):
+    """ Push all content in output_dir to the given GitHub repo branch"""
+    root_tree = _build_git_tree(output_dir, username, password, owner, repo)
+    # get the sha of the latest commit on this branch
+    branch_item = _get_request_content(requests.get,
+                                       '{0}/repos/{1}/{2}/'
+                                       'branches/{3}'.format(_API_URL, owner,
+                                                             repo, branch))
```

Comment:
Optional: <_API_URL>/repos/<owner>/<repo> is common among this file and other utilities.
I recommend to assign it to a variable, like `base_url = '{0}/repos/{1}/{2}/'.format(_API_URL, owner, repo)` at the beginning of this function, and share it among other invocations.